src/Command/Push/PushArtifactCommand.php

```diff
@@ -0,0 +1,242 @@
+    // @todo change deploy strategy depending on whether this is a "source" repo (requiring building) or not
+    // @todo handle if Git user name/email is missing
+    $this->setDirAndRequireProjectCwd($input);
+    $is_dirty = $this->isLocalGitRepoDirty();
+    $commit_hash = $this->getLocalGitCommitHash();
+    if ($is_dirty) {
+      throw new AcquiaCliException('Pushing code was aborted because your local Git repository has uncommitted changes. Please either commit, reset, or stash your changes via git.');
+    }
+    $this->checklist = new Checklist($output);
+
+    // @todo handle environments with tags deployed
+    $this->logger->warning('You must select an environment with a Git branch deployed');
+    $environment = $this->determineEnvironment($input, $output, TRUE);
+    if (strpos($environment->vcs->path, 'tags') === 0) {
+      throw new AcquiaCliException("Environments with Git tags are not supported. Environment {$environment->name} has {$environment->vcs->path} deployed.");
+    }
+    $artifact_dir = Path::join(sys_get_temp_dir(), 'acli-push-artifact');
+    $output_callback = $this->getOutputCallback($output, $this->checklist);
+
+    $this->checklist->addItem('Preparing artifact directory');
+    $this->prepareDir($output_callback, $artifact_dir, $environment->vcs->url, $environment->vcs->path);
+    $this->checklist->completePreviousItem();
+
+    $this->checklist->addItem('Generating build artifact');
+    $this->build($output_callback, $artifact_dir);
+    $this->checklist->completePreviousItem();
+
+    if (!$input->getOption('no-sanitize')) {
+      $this->checklist->addItem('Sanitizing build artifact');
+      $this->sanitize($artifact_dir);
+      $this->checklist->completePreviousItem();
+    }
+
+    $this->checklist->addItem("Committing changes (commit hash: $commit_hash)");
+    $this->commit($output_callback, $artifact_dir, $commit_hash);
+    $this->checklist->completePreviousItem();
+
+    if (!$input->getOption('dry-run')) {
+      $this->checklist->addItem("Pushing changes to {$environment->vcs->path} branch in the {$environment->name} environment");
+      $this->push($output_callback, $artifact_dir);
+      $this->checklist->completePreviousItem();
+    }
+    else {
+      $this->logger->warning("The <options=bold>--dry-run</> option prevented changes from being pushed to Acquia Cloud. The artifact has been built at $artifact_dir");
+    }
+
+    return 0;
+  }
+
+  /**
+   * Prepares a directory to build the artifact.
+   *
+   * @param \Closure $output_callback
+   * @param string $artifact_dir
+   * @param string $vcs_url
+   * @param string $vcs_path
+   *
+   * @throws \Acquia\Cli\Exception\AcquiaCliException
+   */
+  protected function prepareDir(Closure $output_callback, string $artifact_dir, string $vcs_url, string $vcs_path): void {
+    $fs = $this->localMachineHelper->getFilesystem();
+
+    $this->logger->info("Removing $artifact_dir if it exists");
```

Comment:
Use `$this->io`

Reply:
Same with following lines.

Reply:
Actually, it seems like this should be added to the `$output_callback` via `$output_callback('out', "Removing $artifact_dir if it exists")`